drivers/pinctrl/pinctrl_npcx.c

```diff
@@ -0,0 +1,142 @@
+	COND_CODE_1(Z_PINCTRL_NPCX_HAS_IO_PROP(node_id),		\
+		(Z_PINCTRL_NPCX_IO_PAD_CONF_INIT(node_id, io_pad)), ())
+
+#define NPCX_IO_PAD_CONF_LIST						\
+	{DT_FOREACH_CHILD(DT_NODELABEL(pinctrl), NPCX_IO_PAD_CONF_ITEMS)}
+
+const pinctrl_soc_pin_t io_pad_conf[] = NPCX_IO_PAD_CONF_LIST;
+
+static const struct npcx_pinctrl_config npcx_pinctrl_cfg = {
+	.base_scfg = DT_REG_ADDR_BY_NAME(DT_NODELABEL(scfg), scfg),
+};
+
+/* Pin-control local functions for peripheral devices */
+static bool npcx_periph_pinmux_has_lock(int group)
+{
+#if defined(CONFIG_SOC_SERIES_NPCX7)
+	if (group == 0x00 || (group >= 0x02 && group <= 0x04) || group == 0x06 ||
+		group == 0x0b || group == 0x0f) {
+		return true;
+	}
+#elif defined(CONFIG_SOC_SERIES_NPCX9)
+	if (group == 0x00 || (group >= 0x02 && group <= 0x06) || group == 0x0b ||
+		group == 0x0d || (group >= 0x0f && group <= 0x12)) {
+		return true;
+	}
+#endif
+	return false;
+}
+
+static void npcx_periph_pinmux_configure(const struct npcx_periph *alt, int func,
+	bool is_locked)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+	uint8_t alt_mask = BIT(alt->bit);
+
+	/*
+	 * func == 0 means select GPIO, otherwise Alternate Func.
+	 * inverted == 0:
+	 *    Set devalt bit to select Alternate Func.
+	 * inverted == 1:
+	 *    Clear devalt bit to select Alternate Func.
+	 */
+	if (!!func != alt->inverted) {
+		NPCX_DEVALT(scfg_base, alt->group) |=  alt_mask;
+	} else {
+		NPCX_DEVALT(scfg_base, alt->group) &= ~alt_mask;
+	}
+
+	if (is_locked && npcx_periph_pinmux_has_lock(alt->group)) {
+		NPCX_DEVALT_LK(scfg_base, alt->group) |= alt_mask;
+	}
+}
+
+static void npcx_periph_pupd_configure(const struct npcx_periph *pupd,
+	enum npcx_io_bias_type type)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+
+	if (type == NPCX_BIAS_TYPE_NONE) {
+		NPCX_PUPD_EN(scfg_base, pupd->group) &= ~BIT(pupd->bit);
+	} else {
+		NPCX_PUPD_EN(scfg_base, pupd->group) |=  BIT(pupd->bit);
+	}
+}
+
+static void npcx_periph_configure(const pinctrl_soc_pin_t *pin)
+{
+	if (pin->cfg.periph.type == NPCX_PINCTRL_TYPE_PERIPH_PINMUX) {
+		/* Configure peripheral device's pinmux functionality */
+		npcx_periph_pinmux_configure(&pin->cfg.periph, 1,
+					     pin->flags.pinmux_lock);
+	} else if (pin->cfg.periph.type == NPCX_PINCTRL_TYPE_PERIPH_PUPD) {
+		/* Configure peripheral device's internal PU/PD */
+		npcx_periph_pupd_configure(&pin->cfg.periph,
+			pin->flags.io_bias_type);
+	}
+}
+
+/* Pin-control local functions for IO pads */
+static void npcx_iopad_1p8_input_configure(const struct npcx_io_pad *pad)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+
+	NPCX_LV_GPIO_CTL(scfg_base, pad->lvol_ctrl) |= BIT(pad->lvol_bit);
+}
+
+static void npcx_io_pad_configure(const pinctrl_soc_pin_t *pin)
+{
+	if (pin->flags.type == NPCX_PINCTRL_TYPE_IO_PAD) {
+		if (pin->flags.io_1p8v_input) {
+			npcx_iopad_1p8_input_configure(&pin->cfg.io_pad);
+		}
+		/**
+		 * TODO: Add more io-pad configurations here.
+		 */
```

Comment:
maybe return `-ENOTSUP`/`-EINVAL` in the mean time?

Reply:
Done.